cmd/relay-gossip/node/relaynode.go

```diff
@@ -0,0 +1,201 @@
+		return nil, xerrors.Errorf("getting InterfaceListenAddresses: %w", err)
+	}
+
+	for _, a := range addr {
+		log.Infof("%s/p2p/%s\n", a, h.ID())
+	}
+
+	t, err := ps.Join(lp2p.PubSubTopic(cfg.Network))
+	if err != nil {
+		return nil, xerrors.Errorf("joining topic: %w", err)
+	}
+
+	opts := []grpc.DialOption{}
+	if cfg.CertPath != "" {
+		creds, err := credentials.NewClientTLSFromFile(cfg.CertPath, "")
+		if err != nil {
+			return nil, xerrors.Errorf("loading cert file: %w", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else if cfg.Insecure {
+		opts = append(opts, grpc.WithInsecure())
+	} else {
+		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})))
+	}
+	g := &GossipRelayNode{
+		bootstrap: bootstrap,
+		ds:        ds,
+		priv:      priv,
+		h:         h,
+		ps:        ps,
+		t:         t,
+		opts:      opts,
+		addr:      addr,
+		done:      make(chan struct{}),
+	}
+	go g.start(cfg.DrandPublicGRPC)
+	return g, nil
+}
+
+// Addr returns the gossipsub multiaddresses of this relay node.
+func (g *GossipRelayNode) Addr() []ma.Multiaddr {
+	b := make([]ma.Multiaddr, len(g.addr))
+	for i, a := range g.addr {
+		m, err := ma.NewMultiaddr(fmt.Sprintf("%s/p2p/%s", a, g.h.ID()))
+		if err != nil {
+			panic(err)
+		}
+		b[i] = m
+	}
+	return b
+}
+
+// Shutdown stops the relay node.
+func (g *GossipRelayNode) Shutdown() {
+	close(g.done)
+}
+
+func ParseMultiaddrSlice(peer []string) ([]ma.Multiaddr, error) {
+	out := make([]ma.Multiaddr, len(peer))
+	for i, peer := range peer {
+		m, err := ma.NewMultiaddr(peer)
+		if err != nil {
+			return nil, xerrors.Errorf("parsing multiaddr\"%s\": %w", peer, err)
+		}
+		out[i] = m
+	}
+	return out, nil
+}
+
+func (g *GossipRelayNode) start(drandPublicGRPC string) {
+	for {
+		select {
+		case <-g.done:
+			return
+		default:
+		}
+		conn, err := grpc.Dial(drandPublicGRPC, g.opts...)
+		if err != nil {
+			log.Warnf("error connecting to grpc: %+v", err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		client := drand.NewPublicClient(conn)
+		err = g.workRelay(client)
+		if err != nil {
+			log.Warnf("error relaying: %+v", err)
+			time.Sleep(5 * time.Second)
+		}
+	}
+}
+
+func (g *GossipRelayNode) workRelay(client drand.PublicClient) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	curr, err := client.PublicRand(ctx, &drand.PublicRandRequest{Round: 0})
+	if err != nil {
+		return xerrors.Errorf("getting initial round failed: %w", err)
+	}
+	log.Infof("got latest rand: %d", curr.Round)
+
+	// context.Background() on purpose as this applies to whole, long lived stream
+	stream, err := client.PublicRandStream(context.Background(), &drand.PublicRandRequest{Round: curr.Round})
+	if err != nil {
+		return xerrors.Errorf("getting rand stream: %w", err)
+	}
+
+	for {
+		select {
+		case <-g.done:
+			return xerrors.Errorf("relay shutdown")
+		default:
+		}
+		rand, err := stream.Recv()
```

Comment:
I think the shutdown logic and context propagation needs some work - if I'm reading this right I think if we blocking here and get asked to shutdown we have to wait for the next randomness before we'll be able to exit the loop?

Reply:
This is unavoidable due to the stream.Recv() interface.